### PR TITLE
example(dataset): add kitti lane dataset

### PR DESCRIPTION
**Pull request by anda-ren**
## Description
- add kitti lane dataset
- fix self.resp.raw.read() can't read whole file
image

## Modules
- [ ] UI
- [ ] Controller
- [ ] Agent
- [ ] Client
- [ ] Python-SDK
- [ ] Others

## Checklist
- [ ] run code format and lint check
- [ ] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: client/starwhale/core/dataset/store.py [hunk 1]
```diff
@@ -638,7 +638,7 @@ def read(self, size: int) -> bytes:
             return b""
         elif size < 0:
             # urllib3 resp read does not support to read -1
-            return self.resp.raw.read()  # type: ignore
+            return self.resp.content  # type: ignore
```
Comment (tianweidut) [inline client/starwhale/core/dataset/store.py:641]: `self.resp.content = self.resp.raw.read()` and use the uniform code schema for the different sizes (size <0 and size >= 0).

File: client/starwhale/core/dataset/type.py [hunk 2]
```diff
@@ -659,14 +659,20 @@ def __init__(
...
 
     def to_numpy(self) -> numpy.ndarray:
         return numpy.array(self.to_str(), dtype=self.dtype)
 
-    def to_str(self) -> str:
+    def to_str(self, encoding: str = "") -> str:
+        self.link_to_content(encoding)
         return self.content
 
+    def link_to_content(self, encoding: str = "") -> None:
```
Comment (tianweidut) [inline client/starwhale/core/dataset/type.py:672]: move these logical into line649 in the __init__ function? so we can get a unique self.content in the whole Text class.

File: client/starwhale/core/dataset/store.py (hunk 1 above)
Comment (anda-ren) [reply client/starwhale/core/dataset/store.py:641]: sometimes `self.resp.content != self.resp.raw.read()`
```python
>>>import requests
>>> resp=requests.get("https://starwhale-examples.oss-cn-beijing.aliyuncs.com/dataset/kitti/lane/data_road/training/calib/um_000000.txt",stream=True)
>>> b=resp.raw.read()
>>> len(b)
502
>>> resp=requests.get("https://starwhale-examples.oss-cn-beijing.aliyuncs.com/dataset/kitti/lane/data_road/training/calib/um_000000.txt",stream=True)
>>> b=resp.content
>>> len(b)
1860
>>> 

```

File: client/starwhale/core/dataset/type.py (hunk 2 above)
Comment (anda-ren) [reply client/starwhale/core/dataset/type.py:672]: links are used for lazy initialing. If content in link is fetched in init function, the content will be stored into datastore directlly.

Comment (tianweidut) [reply client/starwhale/core/dataset/type.py:672]: ok.
for the lazy loading, we may use the following schema:

```python
from functools import cached_property

def __init__(self, content):
   self._content = content

@cached_property
def content(self):
   if not self._content:
       self._content = self.link_to_content()

   return self._content

...

call `self.content` property 
````

File: client/starwhale/core/dataset/store.py (hunk 1 above)
Comment (tianweidut) [reply client/starwhale/core/dataset/store.py:641]: the streaming mode maybe led the unequal value.

Comment (anda-ren) [reply client/starwhale/core/dataset/store.py:641]: > the streaming mode maybe led the unequal value.

Without streaming mode we can't access `resp.raw.read()` like this. See the doc: https://requests.readthedocs.io/en/latest/user/quickstart/#raw-response-content

> In the rare case that you’d like to get the raw socket response from the server, you can access r.raw. If you want to do this, make sure you set stream=True in your initial request. Once you do, you can do this:
> r = requests.get('https://api.github.com/events', stream=True)
> r.raw
> <urllib3.response.HTTPResponse object at 0x101194810>
> r.raw.read(10)
> '\x1f\x8b\x08\x00\x00\x00\x00\x00\x00\x03'

File: client/starwhale/core/dataset/type.py (hunk 2 above)
Comment (anda-ren) [reply client/starwhale/core/dataset/type.py:672]: > got.
> 
> for the lazy loading, we may use the following schema:
> 
> ```python
> from functools import cached_property
> 
> def __init__(self, content):
>    self._content = content
> 
> @cached_property
> def content(self):
>    if not self._content:
>        self._content = self.link_to_content()
> 
>    return self._content
> 
> ...
> 
> call `self.content` property in `to_bytes` or `to_str` function.
> ```

OK

Comment (anda-ren) [reply client/starwhale/core/dataset/type.py]: done

**Review by tianweidut — APPROVED** [review]
LGTM

**Pull request merged**